docs/docs/reconciliation.md

```diff
@@ -0,0 +1,123 @@
+
+```xml
+<div className="before" title="stuff" />
+
+<div className="after" title="stuff" />
+```
+
+When converting between these two elements, React knows to only modify the `className`.
+
+When updating `style`, React also knows to update only the properties that changed. For example:
+
+```xml
+<div style={{'{{'}}color: 'red', fontWeight: 'bold'}} />
+
+<div style={{'{{'}}color: 'green', fontWeight: 'bold'}} />
+```
+
+When converting between these two elements, React knows to only modify the `color` style, not the `fontWeight`.
+
+After handling the DOM element, React then recurses on the children.
+
+### Custom Components
+
+When a component updates, the instance stays the same, so that state is maintained across renders. React takes all the attributes from the new component element and calls `componentWillReceiveProps()` and `componentWillUpdate()` on the previous one.
+
+Next, the `render()` method is called and the diff algorithm recurses on the previous result and the new result.
+
+### Recursing On Children
+
+By default, when recursing on the children of a DOM element, React just iterates over both lists of children at the same time and generates a mutation whenever there's a difference.
+
+For example, when adding an element at the end of the children, converting between these two trees works well:
+
+```xml
+<div>
+  <span>first</span>
+  <span>second</span>
+</div>
+
+<div>
+  <span>first</span>
+  <span>second</span>
+  <span>third</span>
+</div>
+```
+
+React will match the two `<span>first</span>` trees, match the two `<span>second</span>` trees, and then insert the `<span>third</span>` tree.
+
+Inserting an element at the beginning has worse performance. For example, converting between these two trees works poorly:
+
+```xml
+<div>
+  <span>Duke</span>
+  <span>Villanova</span>
+</div>
+
+<div>
+  <span>Connecticut</span>
+  <span>Duke</span>
+  <span>Villanova</span>
+</div>
+```
+
+React will mutate every child instead of realizing it can keep the `<span>Duke</span>` and `<span>Villanova</span>` subtrees intact. This inefficiency can be a problem.
+
+### Keys
+
+In order to solve this issue, React supports an optional `key` attribute. When children have keys, React uses the key to match children in the original tree with children in the subsequent tree. For example, adding a `key` to our inefficient example above can make the tree conversion efficient:
+
+```xml
+<div>
+  <span key={2015}>Duke</span>
+  <span key={2016}>Villanova</span>
+</div>
+
+<div>
+  <span key={2014}>Connecticut</span>
+  <span key={2015}>Duke</span>
+  <span key={2016}>Villanova</span>
+</div>
+```
+
+In practice, finding a key is not really hard. Most of the time, the element you are going to display already has a unique id. When that's not the case, you can add a new ID property to your model or hash some parts of the content to generate a key. The key only has to be unique among its siblings, not globally unique.
```

Comment:
Maybe add "As a last resort, you can pass item's index in the array as a key. This can work well if the items are never re-ordered, but reorders will be slow."